src/flag_gems/ops/isin.py

```diff
@@ -0,0 +1,294 @@
+        )
+    else:  # grid-stride-loop style kernel
+        for j in range(0, tiles_per_cta):
+            global_pid = pid + j * num_ctas if j > 0 else pid
+            isin_by_search_impl(
+                global_pid,
+                in0_ravel_ptr,
+                in1_sorted_ptr,  # in
+                out_ptr,  # out
+                M,
+                N,
+                BLOCK_M,
+                invert,
+            )
+
+
+def isin_by_search(
+    in0: torch.tensor,
+    in1: torch.tensor,
+    invert: bool,
+    unique_in0: bool,
+    unique_in1: bool,
+):
+    # unique or sort or ravel
+    if unique_in0:
+        in0_ravel, unique_order, _ = _unique2(
+            in0, sorted=True, return_inverse=True, return_counts=False
+        )
+    else:
+        in0_ravel = in0.contiguous().ravel()
+    if unique_in1:
+        in1_ravel, _, _ = _unique2(
+            in1, sorted=True, return_inverse=False, return_counts=False
+        )
+    else:
+        in1_ravel, _ = torch.sort(in1.ravel())
+    # launch kernel func
+    M = in0_ravel.numel()
+    N = in1_ravel.numel()
+    if M <= 1048576:  # 2 ** 20 = 1024 * 1024
+        _, BLOCK_M, num_warps = launch_arg(None, 512, M, 8)
+    elif M <= 4194304:  # 2 ** 22 = 1024 * 4096
+        _, BLOCK_M, num_warps = launch_arg(None, 1024, M, 8)
+    elif M <= 8388608:  # 2 ** 23 = 1024 * 8192
+        _, BLOCK_M, num_warps = launch_arg(None, 2048, M, 16)
+    elif M <= 268435456:  # 2 ** 28 = 1024 * 262144
+        _, BLOCK_M, num_warps = launch_arg(None, 4096, M, 32)
+    else:
+        _, BLOCK_M, num_warps = launch_arg(None, 2048, M, 16)
+    num_ctas = min(65536, triton.cdiv(M, BLOCK_M))
+    tiles_per_cta = triton.cdiv(M, BLOCK_M * num_ctas)
+    grid = (num_ctas,)
+    out = torch.empty_like(in0_ravel, dtype=torch.bool)
+    with torch.cuda.device(in0_ravel.device.index):
+        isin_by_search_kernel[grid](
+            in0_ravel,
+            in1_ravel,  # in
+            out,  # out
+            M,
+            N,
+            BLOCK_M,
+            tiles_per_cta=tiles_per_cta,
+            one_tile_per_cta=tiles_per_cta == 1,
+            invert=invert,
+            num_warps=num_warps,
+        )
+    if unique_in0:
+        out = torch.gather(out, 0, unique_order.ravel().to(torch.int64))
+    return out.view_as(in0)
+
+
+def isin(
+    in0,
+    in1,
+    *,
+    assume_unique: bool = False,
+    invert: bool = False,
+) -> torch.Tensor:
+    if not torch.is_tensor(in0):
+        assert torch.is_tensor(in1)
+        in0 = torch.tensor(in0, device=in1.device)
+    elif not torch.is_tensor(in1):
+        assert torch.is_tensor(in0)
+        in1 = torch.tensor(in1, device=in0.device)
+    if in0.numel() == 0 or in1.numel() == 0:
+        return torch.zeros_like(in0, dtype=torch.bool)
+    elif in0.numel() <= 12288 and in1.numel() <= 12288:  # 1024 * 12
+        return isin_by_comparation(in0, in1, invert)
+    elif assume_unique or in1.numel() <= 4194304:  # 1024 * 4096
```

Comment:
Here why use 1024 * 12 and 1024 * 4096 as the condition ?

Reply:
> Here why use `1024 * 12` and `1024 * 4096` as the condition ?

It's the autotune result.